src/django_idom/hooks.py

```diff
@@ -30,3 +39,145 @@ def use_websocket() -> IdomWebsocket:
+
+    data, set_data = use_state(UNDEFINED)
+    loading, set_loading = use_state(True)
+    error, set_error = use_state(None)
+
+    @use_callback
+    def refetch() -> None:
+        set_data(UNDEFINED)
+        set_loading(True)
+        set_error(None)
+
+    @use_effect(dependencies=[])
+    def add_refetch_callback():
+        # By tracking callbacks globally, any usage of the query function will be re-run
+        # if the user has told a mutation to refetch it.
+        _REFETCH_CALLBACKS[query].add(refetch)
+        return lambda: _REFETCH_CALLBACKS[query].remove(refetch)
+
+    @use_effect(dependencies=None)
+    def execute_query():
+        if data is not UNDEFINED:
+            return
+
+        def thread_target():
+            try:
+                query_result = query(*args, **kwargs)
+            except Exception as e:
+                set_data(UNDEFINED)
+                set_loading(False)
+                set_error(e)
+                return
+
+            if isinstance(query_result, QuerySet):
+                if fetch_deferred_fields:
+                    for model in query_result:
+                        _fetch_deferred_fields(model)
+                else:
+                    # still force query set to execute
+                    for _ in query_result:
+                        pass
+            elif isinstance(query_result, Model):
+                if fetch_deferred_fields:
+                    _fetch_deferred_fields(query_result)
+            elif fetch_deferred_fields:
+                raise ValueError(
+                    f"Expected {query} to return Model or Query because "
+                    f"{fetch_deferred_fields=}, got {query_result!r}"
+                )
+
+            set_data(query_result)
+            set_loading(False)
+            set_error(None)
+
+        # We need to run this in a thread so we don't prevent rendering when loading.
+        # We also can't do this async since Django's ORM doesn't support this yet.
+        Thread(target=thread_target, daemon=True).start()
+
+    return Query(data, loading, error, refetch)
+
+
+@dataclass
+class Query(Generic[_Data]):
+    data: _Data
+    loading: bool
+    error: Exception | None
+    refetch: Callable[[], None]
+
+
+def use_mutation(
+    mutate: Callable[_Params, None],
+    refetch: Callable[..., Any] | Sequence[Callable[..., Any]],
+) -> Mutation[_Params]:
+    loading, set_loading = use_state(True)
+    error, set_error = use_state(None)
+
+    @use_callback
+    def call(*args: _Params.args, **kwargs: _Params.kwargs) -> None:
+        set_loading(True)
+
+        def thread_target():
+            try:
+                mutate(*args, **kwargs)
+            except Exception as e:
+                set_loading(False)
+                set_error(e)
+            else:
+                set_loading(False)
+                set_error(None)
+                for query in (refetch,) if isinstance(refetch, Query) else refetch:
```

Comment:
I incorrectly implemented this check. It should actually be `isinstance(refetch, FunctionType)`